test/celeritas/track/TrackSort.test.cc

```diff
@@ -0,0 +1,179 @@
+        }
+        return result;
+    }
+
+    //! Make 10GeV electrons along +x
+    std::vector<Primary> make_primaries(size_type count) const override
+    {
+        return this->make_primaries_with_energy(count, MevEnergy{10000});
+    }
+
+    size_type max_average_steps() const override
+    {
+        return 100000;  // 8 primaries -> ~500k steps, be conservative
+    }
+
+    GeantPhysicsOptions build_geant_options() const override
+    {
+        auto opts = TestEm3Base::build_geant_options();
+        opts.msc = MscModelSelection::none;
+        return opts;
+    }
+};
+
+TEST_F(TestTrackPartitionEm3Stepper, host_is_partitioned)
+{
+    size_type num_primaries = 8;
+    size_type num_tracks = 128;
+
+    Stepper<MemSpace::host> step(this->make_stepper_input(num_tracks));
+
+    // Initialize some primaries and take a step
+    auto primaries = this->make_primaries(num_primaries);
+    step(make_span(primaries));
+    auto check_is_partitioned = [&step]() {
+        auto span = step.core_data()
+                        .states.track_slots[AllItems<TrackSlotId::size_type>{}];
+        return std::is_partitioned(
+            span.begin(),
+            span.end(),
+            [&status = step.core_data().states.sim.status](
+                auto const track_slot) {
+                return status[TrackSlotId{track_slot}] == TrackStatus::alive;
+            });
+    };
+    // we partition at the start of the step so we need to explictly partition
+    // again after a step before checking
+    for (auto i = 0; i < 10; ++i)
+    {
+        detail::partition_tracks_by_status(step.core_data().states);
+        EXPECT_TRUE(check_is_partitioned()) << "Track slots are not "
+                                               "partitioned by status";
+        step();
+    }
+    step(make_span(primaries));
+    for (auto i = 0; i < 10; ++i)
+    {
+        detail::partition_tracks_by_status(step.core_data().states);
+        EXPECT_TRUE(check_is_partitioned()) << "Track slots are not "
+                                               "partitioned by status";
+        step();
+    }
+}
+
+TEST_F(TestTrackPartitionEm3Stepper,
+       TEST_IF_CELER_DEVICE(device_is_partitioned))
+{
+    size_type num_primaries = 8;
+    // Num tracks is low enough to hit capacity
+    size_type num_tracks = num_primaries * 800;
+
+    Stepper<MemSpace::device> step(this->make_stepper_input(num_tracks));
+
+    // Initialize some primaries and take a step
+    auto primaries = this->make_primaries(num_primaries);
+    step(make_span(primaries));
+    auto check_is_partitioned = [&step]() {
+        // copy to host
+        auto core_ref = step.core_data();
+        Collection<TrackSlotId::size_type, Ownership::value, MemSpace::host, ThreadId>
+            track_slots;
+        track_slots = core_ref.states.track_slots;
+        StateCollection<TrackStatus, Ownership::value, MemSpace::host> track_status;
+        track_status = core_ref.states.sim.status;
```

Comment:
Has to be done in two statement otherwise it'll try to convert `track_slots` and use copy initialization instead of assignment operator. Not sure if there is a cleaner way of doing DtoH memory transfer.

Reply:
I think it's all we can do for now. If we want something fancy to help with unit tests, we could define a helper function...